runtime/lua/vim/lsp/buf.lua

```diff
@@ -143,6 +143,82 @@ local function select_client(method, on_choice)
   end
 end
 
+--- Formats a buffer using the attached (and optionally filtered) language
+--- server clients.
+---
+--- @param opts table|nil Optional table which holds the following optional fields:
+---     - formatting_options (table|nil):
+---         Can be used to specify FormattingOptions. Some unspecified options will be
+---         automatically derived from the current Neovim options.
+---         @see https://microsoft.github.io/language-server-protocol/specification#textDocument_formatting
+---     - timeout_ms (integer|nil, default 1000):
+---         Time in milliseconds to block for formatting requests. Formatting requests are current
+---         synchronous to prevent editing of the buffer.
+---     - bufnr (number|nil):
+---         Restrict formatting to the clients attached to the given buffer, defaults to the current
+---         buffer (0).
+---     - filter (function|nil):
+---         Predicate to filter clients used for formatting. Received a client
+---         as argument and must return a boolean.
+---
+---         Example:
+---         <pre>
+---         -- Never request typescript-language-server for formatting
+---         vim.lsp.buf.format {
+---           filter = function(clients)
+---             return vim.tbl_filter(
+---               function(client) return client.name ~= "tsserver" end,
+---               clients
+---             )
+---           end
+---         }
+---         </pre>
+---
+---     - id (number|nil):
+---         Restrict formatting to the client with ID (client.id) matching this field.
+---     - name (string|nil):
+---         Restrict formatting to the client with name (client.name) matching this field.
+
+function M.format(opts)
+  opts = opts or {}
+  local bufnr = opts.bufnr or vim.api.nvim_get_current_buf()
+  local clients = vim.lsp.buf_get_clients(bufnr)
+
+  if opts.filter then
+    clients = opts.filter(clients)
+  elseif opts.id then
+    clients = vim.tbl_filter(
+      function(client) return client.id == opts.id end,
+      clients
+    )
+  elseif opts.name then
+    clients = vim.tbl_filter(
+      function(client) return client.name == opts.name end,
+      clients
+    )
+  end
+
+  clients = vim.tbl_filter(
+    function(client) return client.supports_method("textDocument/formatting") end,
+    clients
+  )
+
+  if #clients == 0 then
+    vim.notify("[LSP] Format request failed, no matching language servers.")
+  end
+
+  local timeout_ms =  opts.timeout_ms or 1000
+  for _, client in pairs(clients) do
+      local params = util.make_formatting_params(opts.formatting_options)
+      local result, err = client.request_sync("textDocument/formatting", params, timeout_ms, bufnr)
```

Comment:
Any thoughts on adding an `async` option?

Reply:
This was a less dramatic attempt at https://github.com/neovim/neovim/pull/16206, I don't want to have a sync vs. async distinction. I would like to have chained async that blocks the neovim shutdown process so we can always use async, but lock the buffer while pending requests are out to prevent clobbering the buffer with wrong text edits (same issue currently with trying to request multiple servers asynvc